Feature multi-model

IntegralProgrammer opened the pull request:
Extends the API so that:
  - A trained model can be selected for an annotation request (eg. HPO, MONDO, etc...)
  - Trained models can be listed (`/lsmodels`)
  - Trained models can be deleted (`/delete_model`)
  - A script can be generated for model training on HPF (`/submit_training_job`)

Modifications were made to the model training code so that it executes on HPF.

IntegralProgrammer committed:
Fix incorrect syntax for print() statement

Redefine print() statement so that log entries are written to a file (not STDOUT) - needed as a workaround for a bug on HPF

IntegralProgrammer committed:
Add multi-model API capable of generating training jobs for HPF

IntegralProgrammer committed:
Update README.md with multi-model example

sdumitriu commented on docker_run_webapp.sh:
```diff
@@ -0,0 +1,14 @@
+	-v $(realpath ~/ncr_qsub):/root/qsub:rw \
+	-v $(realpath ~/ncr_model_params):/root/opt/ncr/model_params:ro \
+	-p 127.0.0.1:5000:5000 \
+	-e AUTOTEST=$AUTOTEST \
```
This belongs to a different PR.

IntegralProgrammer replied:
Done - this content has been removed and will be later added in a different PR :heavy_check_mark:

sdumitriu commented on docker_run_webapp.sh:
```diff
@@ -0,0 +1,14 @@
+	-v $(realpath ~/ncr_model_params):/root/opt/ncr/model_params:ro \
+	-p 127.0.0.1:5000:5000 \
+	-e AUTOTEST=$AUTOTEST \
+	-e TEST_IGNORE_SCORE=$TEST_IGNORE_SCORE \
```
This too.

IntegralProgrammer replied:
Done - this content has been removed and will be later added in a different PR :heavy_check_mark:

sdumitriu commented on app.py:
```diff
@@ -1,20 +1,47 @@
+#Stored in a form of {"object": model object, "threshold": threshold value}
+NCR_MODELS = {}
 
+#For now, to be later fixed...
```
When later?

IntegralProgrammer replied:
I revised this comment clarifying its meaning - see line 18 in `app.py`. In this early version of the multi-model API, models are to be added by modification of the lines below defining the object `NCR_MODELS`. In upcoming Pull Requests, API endpoints will be created for model training, and after successful model training, these models will be automatically added to the `NCR_MODELS` object.

sdumitriu commented on app.py:
```diff
@@ -49,6 +76,68 @@ def dated_url_for(endpoint, **values):
     return url_for(endpoint, **values)
 
 
+@app.route('/lsmodels/')
```
This isn't very RESTful, where paths should be the name of resources, and the HTTP commands should be actions.

So, ideally `GET /models` should list the existing models, `POST /models` (or maybe `PUT /models/HPO`?) should submit a training job, and `DELETE /models/HPO` should delete one model. Can Flask do that easily, route different HTTP methods to different Python methods?

IntegralProgrammer replied:
I have made revisions, you can now `GET /models/` to see the list of available trained models and you can also `DELETE /models/NameOfModelToDelete` to remove a model. Routing different HTTP methods to different Python methods in Flask is simple. `PUT` / `POST` functionality will be added in the PR containing the model training code.

sdumitriu commented on app.py:
```diff
@@ -284,12 +377,16 @@ def annotate_post():
 @app.route('/annotate/', methods=['GET'])
 def annotate_get():
+    if not 'model' in request.args:
+        abort(400)
```
I would add some backwards compatibility for this: if there's only one model trained, and there's no `model` specified in the request, then just use the one and only model available.

IntegralProgrammer replied:
Done - I have added the backwards compatibility as you have described. :heavy_check_mark:

sdumitriu reviewed (changes requested):
I would include only the changes needed for multiple models in this PR, moving the HPF integration in a different PR.

IntegralProgrammer committed:
Revise according to PR feedback.

Extend multi-model support to POST methods in addition to GET methods

IntegralProgrammer committed:
Revise Dockerfile

According to PR feedback, model training/HPF integration should be in a different PR.

IntegralProgrammer committed:
Removal of HPF/Model Training code from this PR

IntegralProgrammer commented:
Thank you for your feedback. I have revised this PR so that it only includes the changes needed for using multiple models and does not include any model training or HPF code.
